cpp/src/arrow/record_batch.cc

```diff
@@ -623,6 +667,16 @@ Status RecordBatch::ValidateFull() const {
   return ValidateBatch(*this, /*full_validation=*/true);
 }
 
+const std::shared_ptr<Device::SyncEvent>& RecordBatch::GetSyncEvent() const {
+  return null_sync_event_;
+}
+
+DeviceAllocationType RecordBatch::device_type() const {
+  return DeviceAllocationType::kCPU;
```

Comment:
Why is this not returning `device_type_`?

Reply:
because `device_type_` was added to the `SimpleRecordBatch` derived class from RecordBatch, it's not in the base class. I didn't want to force implementations to do it that way if they potentially could pull the device_type from elsewhere.

So the default implementation just returns CPU, but @bkietz brought up a good point https://github.com/apache/arrow/pull/40807#discussion_r1586298099 where it might make sense to just make this pure virtual instead of having a default impl

Reply:
Ah, I missed the distinction between RecordBatch and SimpleRecordBatch in the diff